crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -48,430 +231,413 @@ where
     }
 }
 
-impl<Client, Consensus> BodyDownloader for ConcurrentDownloader<Client, Consensus>
+impl<B, C, DB> BodyDownloader for ConcurrentDownloader<B, C, DB>
 where
-    Client: BodiesClient,
-    Consensus: ConsensusTrait,
+    B: BodiesClient + 'static,
+    C: ConsensusTrait + 'static,
+    DB: Database,
 {
-    fn bodies_stream<'a, 'b, I>(&'a self, headers: I) -> DownloadStream<'a, BlockResponse>
-    where
-        I: IntoIterator<Item = &'b SealedHeader>,
-        <I as IntoIterator>::IntoIter: Send + 'b,
-        'b: 'a,
-    {
-        Box::pin(
-            stream::iter(headers.into_iter())
-                .chunks(self.batch_size)
-                .map(move |headers| {
-                    (move || self.fetch_bodies(headers.clone()))
-                        // We should not backoff from requests since the downloader does
-                        // not know what peer the request is being delivered to.
-                        // See: https://github.com/paradigmxyz/reth/issues/809
-                        .retry(
-                            ConstantBackoff::default()
-                                .with_delay(Duration::from_secs(0))
-                                .with_max_times(self.retries),
-                        )
-                })
-                .buffered(self.concurrency)
-                .map_ok(|blocks| stream::iter(blocks.into_iter()).map(Ok))
-                .try_flatten(),
-        )
+    /// Set a new download range (exclusive).
+    ///
+    /// This method will drain all queued bodies and re-request any
+    /// missing blocks with number less than the last queued block.
+    fn set_download_range(&mut self, range: Range<BlockNumber>) -> Result<(), db::Error> {
+        if range.is_empty() {
+            tracing::warn!(target: "downloaders::bodies", "New header range is empty");
+            return Ok(())
+        }
+
+        tracing::trace!(target: "downloaders::bodies", ?range, "Setting new header range");
+
+        // Drain queued bodies.
+        let queued_bodies = std::mem::take(&mut self.queued_bodies)
+            .into_iter()
+            .filter(|b| range.contains(&b.block_number()))
+            .collect::<Vec<_>>();
```

Comment:
The queued bodies don't go back into the queue, they go back into the buffer. This might answer a couple of comments below as well

Reply:
> The queued bodies don't go back into the queue, they go back into the buffer.

but why?

Reply:
because the queue contains the ordered contiguous block responses that must begin at the new range start. if the stage is resetting the download range to the same as before because of some db error, the dl must first download the bodies from range start and put them into the queue

Reply:
tbh this doesn't make much sense to me atm, but let's go over this later.